Partial fix to attaching VS Code

goldenstein64 opened the pull request:
Changes:

* `RobloxStudioWindow.cs` has been updated to work with Roblox Studio's updated `Automation` structure.
  * Removed public field `EditorParent` for being irrelevant and not being used outside of this module.

Existing bugs:

* VS Code might not overlay over the correct script editor in the *very* specific situation where two scripts of the same name are open and one is in a detached window (in a dock state other than inside the main view). In that case, VS Code will always overlay onto the detached window.
  * This is because I couldn't find a way to differentiate between whether the active script is in the main view or the detached window. It will always choose the detached window in this situation.
  * The `HasKeyboardFocus` property can't be used to determine this anymore because it now points to an instance in the Explorer widget for some reason.
  * This doesn't happen at all for scripts of the same name that are either both in the main view or both in a detached window. Main window editors always occupy the same space, and floating window editors are ordered such that the top-most and coincidentally focused window is searched first.

Personally, I don't ever use detached script editors, so I could use this program as-is, but it may be annoying for other people if they use detached script editors often.

cc #3

goldenstein64 committed:
Fixed Studio Window misreference

I corrected how the RobloxStudioWindow finds the editor window, but the repositioning/resizing of the editor still needs to be fixed.

My first guess is using the one of the parent elements' positions/sizes in place of the old one.

goldenstein64 committed:
Fixed main window resizing

This is very close to being finished - all that's left is fixing floating windows!

goldenstein64 committed:
Fixed floating window focus bug, sort of

There's an issue with the selected instance not matching up with the editor window. Although the selected instance is the only element with keyboard focus, it has some weird behavior. I would have to select the instance in the explorer, and the overlay would conform to that instance's open script editor.

I should probably see if there are other sources for the name of the script other than the element with keyboard focus.

goldenstein64 committed:
Changed how the script name was retrieved

I suppose since we have to open the script to run it, we can store the name of the script there and use it to find our script editor window.

So this marks everything I wanted to solve with this program. I might refactor this on another pull reguest, I'm not sure.

goldenstein64 committed:
Delete test.rbxlx.lock

goldenstein64 committed:
Delete test.rbxlx

goldenstein64 committed:
reverted floating window search, added/changed private methods

TheNexusAvenger commented:
As a heads up, I am acknowledging that this pull request exists, but I will be delayed looking over your comments (cool to see how much you are sharing) and test the code. You may have noticed some of the code before did a lot more traversal than you may expect, and that was because some systems were *very* slow last time I tried this. Maybe this code doesn't have this problem; I haven't profiled it yet.

goldenstein64 commented:
Honestly, I find it extraordinarily hard to understand the `Automation` tree. I could try frontloading element traversal to the constructor once I figure out how to filter out elements more narrowly. My current plan is to build a `DebugElement` class that wraps an `AutomationElement` and has a `Children` field so I can build an Explorer-esque environment through the watch widget and set a handy breakpoint on the line after.

I have a feeling this code will be slower; I can't be sure since I never used NEE before it broke.

goldenstein64 committed:
Added stricter AutomationElement traversal

I also added a DebugElement class for viewing the AutomationElement tree.

goldenstein64 committed:
Added some helpful comments

goldenstein64 commented:
I figured out exactly what paths I need to traverse to get to any of the script editors, but I think I'll only be able to frontload element traversal for the main editor. Detached editor windows are just one level below the root and I'm assuming they can be removed at any time, so I would need to traverse back down to the editor every time the window is created.

TheNexusAvenger commented on NexusEmbeddedEditor/NexusEmbeddedEditor/Window/RobloxStudioWindow.cs:
```diff
@@ -16,18 +16,72 @@
+     * There should now be a tree in the Locals widget containing every element.
+     */
+    /*
+    public class DebugElement
```
I am not a fan of having this code commented out with the intent that developers will un-comment it out for testing. Undoing it may be overlooked. Can using this be a command-line argument option?

goldenstein64 replied:
Would excluding it with an `#if` directive be alright?

```c#
namespace NexusEmbeddedEditor.Window
{
#if DEBUG
    /*
     * doc
     */
    public class DebugElement
    {
        // code
    }
#endif

    // ...
```

This removes it from the build if the `Debug` configuration in the top ribbon bar is changed to `Release`, causing compile errors so you don't accidentally reference `DebugElement` in production.

Reference: https://stackoverflow.com/questions/337431/removing-code-from-release-build-in-net

I'll add a commit in advance.

TheNexusAvenger replied:
That would be better. Expanding the `#if` usage for the actual wrapping call would be better since it would reduce the risk of accidentally committing debug code. I say that as someone who hasn't used the debug code though.

TheNexusAvenger reviewed (changes requested):
Apologies for not reviewing this sooner. Still haven't looked at the functionality because there is 1 change I don't want to be merged.

goldenstein64 committed:
DebugElement is only built in Debug configuration

goldenstein64 committed:
Added some exception handlers

During some stress testing (or just moving the editor between the main view and a detached window quickly), the program would error from elements not existing anymore because I assume the user could tamper with the AutomationElement tree mid-traversal. I added some try-catch statements to stop the error.

goldenstein64 committed:
Merge branch 'TheNexusAvenger:master' into master

TheNexusAvenger reviewed (approved):
Finally ran the code and the functionality seems fine and I am going to keep my last comment as resolved. I would prefer `DebugElement` to be used with `#if DEBUG`, but that change seems much more involved since it doesn't extend `AutomationElement`. I will merge the changes and make the release after you confirm it is done.

goldenstein64 committed:
Added main view caching

Instead of trying to frontload the window on initialization, I decided to try some simple caching when calling GetMainViewEditorWindow(). I assume it would at least be marginally faster than performing the root traversal every time.

goldenstein64 commented:
Now that I added a caching mechanism to the main view window, I think I am finished contributing to the pull request once this last commit is approved.
